Add batching to `IterableDataset`

lappemic opened the pull request:
I've taken a try at implementing a batched `IterableDataset` as requested in issue #6279. This PR adds a new `BatchedExamplesIterable` class and a `.batch()` method to the `IterableDataset` class.

The main changes are:
1. A new `BatchedExamplesIterable` that groups examples into batches.
2. A `.batch()` method for `IterableDataset` to easily create batched versions.
3. Support for shuffling and sharding to work with PyTorch DataLoader and multiple workers.

I'm not sure if this is exactly what you had in mind and also have not fully tested it atm, so I'd really appreciate your feedback. Does this seem like it's heading in the right direction? I'm happy to make any changes or explore different approaches if needed.

Pinging @lhoestq

lhoestq commented:
Cool ! Thanks for diving into it :)

Your implementation is great for and indeed supports shuffling and batching, you just need to additionally account for state_dict (for dataset [checkpointing+resuming](https://huggingface.co/docs/datasets/main/en/use_with_pytorch#checkpoint-and-resume))

That being said, I believe the implementation can be made simpler by relying on `IterableDataset.map()` which already implements all this. Maybe something like

```python

def batch(self, batch_size: int, drop_last_batch: bool = False) -> "IterableDataset":
    def batch(unbatched: dict[str, list]) -> dict[str, list]:
        return {k: [v] for k, v in unbatched}

    return self.map(batch, batched=True, batch_size=batch_size, drop_last_batch=drop_last_batch)
```

And this way no need to reimplement everything !

(my only small concern is that it's not an Arrow-optimized function so it requires the examples to be manipulated as python objects even if the original data is in Arrow format (e.g. when streaming Parquet files) but it's not a big deal and we can see later if we need to optimize this)

lappemic commented:
Thanks a lot for the feedback @lhoestq! I definitely could have saved some time looking into it properly first. 😅 

Implemented the `.batch()` method, added a proper docsrtring for documentation, and added tests.

Let me know what you think and if this needs some update.

HuggingFaceDocBuilderDev commented:
The docs for this PR live [here](https://moon-ci-docs.huggingface.co/docs/datasets/pr_7054). All of your documentation changes will be reflected on that endpoint. The docs are available until 30 days after the last update.

lhoestq commented on src/datasets/iterable_dataset.py:
```diff
@@ -2885,6 +2885,28 @@ def _resolve_features(self):
+        Group samples from the dataset into batches.
+
+        Args:
+            batch_size (int):
+                The number of samples in each batch.
+            drop_last_batch (bool, defaults to `False`):
+                Whether to drop the last incomplete batch.
```
```suggestion
            batch_size (`int`): The number of samples in each batch.
            drop_last_batch (`bool`, defaults to `False`): Whether to drop the last incomplete batch.
```

lhoestq commented on docs/source/about_mapstyle_vs_iterable.mdx:
````diff
@@ -205,6 +205,43 @@ for epoch in range(n_epochs):
+batched_dataset = dataset.batch(batch_size=32, drop_last_batch=True)
+```
+
+Batching with `IterableDataset` combines the benefits of processing data in larger chunks with the memory efficiency of streaming data. 
+This makes it an excellent choice for working with large datasets that don't fit in memory, especially when used with deep learning frameworks that expect batched inputs.
+
````
feels a bit redundant

```suggestion
```

lhoestq commented on docs/source/about_mapstyle_vs_iterable.mdx:
````diff
@@ -205,6 +205,43 @@ for epoch in range(n_epochs):
         pass
 ```
 
+## Batching data in IterableDataset
````
this can go in stream.mdx instead ?

Btw in the future it would be cool to have `Dataset.batch()` as well :)

lhoestq reviewed (commented):
Nice thanks ! I added a few suggestions before we merge.

lappemic commented:
Thanks for the feedbak @lhoestq!

Applied it and referenced the `batched=True` option in the `map` function and highlighted the difference. Hope i got this right.

lhoestq reviewed (approved):
lgtm !
